src/poc/miner_poc_mgr.erl

```diff
@@ -0,0 +1,1118 @@
+                            {noreply, State};
+                        false ->
+                            lager:warning("Got unexpected receipt from ~p for onionkeyhash", [Gateway, OnionKeyHash]),
+                            {noreply, State}
+                    end
+            end
+    end.
+
+%% ------------------------------------------------------------------
+%% Internal functions
+%% ------------------------------------------------------------------
+initialize_poc(BlockHash, POCStartHeight, Keys, Vars, #state{chain = Chain, pub_key = Challenger} = State) ->
+    Ledger = blockchain:ledger(Chain),
+    #{public := OnionCompactKey, secret := {ecc_compact, POCPrivKey}} = Keys,
+    POCPubKeyBin = libp2p_crypto:pubkey_to_bin(OnionCompactKey),
+    #'ECPrivateKey'{privateKey = PrivKeyBin} = POCPrivKey,
+    POCPrivKeyHash = crypto:hash(sha256, PrivKeyBin),
+    OnionKeyHash = crypto:hash(sha256, POCPubKeyBin),
+    Entropy = <<OnionKeyHash/binary, BlockHash/binary>>,
+    ZoneRandState = blockchain_utils:rand_state(Entropy),
+    InitTargetRandState = blockchain_utils:rand_state(POCPrivKeyHash),
+    TargetMod = blockchain_utils:target_v_to_mod(blockchain:config(?poc_targeting_version, Ledger)),
+    case TargetMod:target(Challenger, InitTargetRandState, ZoneRandState, Ledger, Vars) of
+        {error, Reason}->
+            lager:info("failed to find a target for poc key ~p, reason ~p", [OnionKeyHash, Reason]),
+            noop;
+        {ok, {TargetPubkeybin, TargetRandState}}->
+            {ok, LastChallenge} = blockchain_ledger_v1:current_height(Ledger),
+            {ok, B} = blockchain:get_block(LastChallenge, Chain),
+            Time = blockchain_block:time(B),
+            Path = blockchain_poc_path_v4:build(TargetPubkeybin, TargetRandState, Ledger, Time, Vars),
+            N = erlang:length(Path),
+            [<<IV:16/integer-unsigned-little, _/binary>> | LayerData] = blockchain_txn_poc_receipts_v2:create_secret_hash(
+                Entropy,
+                N + 1
+            ),
+            OnionList = lists:zip([libp2p_crypto:bin_to_pubkey(P) || P <- Path], LayerData),
+            {Onion, Layers} = blockchain_poc_packet_v2:build(Keys, IV, OnionList),
+            [_|LayerHashes] = [crypto:hash(sha256, L) || L <- Layers],
+            Challengees = lists:zip(Path, LayerData),
+            PacketHashes = lists:zip(Path, LayerHashes),
+            Secret = libp2p_crypto:keys_to_bin(Keys),
+            %% save the POC data to our local cache
+            LocalPOC = #local_poc{
+                onion_key_hash = OnionKeyHash,
+                block_hash = BlockHash,
+                target = TargetPubkeybin,
+                onion = Onion,
+                secret = Secret,
+                challengees = Challengees,
+                packet_hashes = PacketHashes,
+                keys = Keys,
+                start_height = POCStartHeight
+            },
+            ok = write_local_poc(LocalPOC, State),
```

Comment:
Check if the first challengee is in the denylis here, and if it is return `noop` instead?